rofl-scheduler/src/serverd/auth.rs

```diff
@@ -0,0 +1,228 @@
+        )
+        .map_err(|_| Error::BadAuthToken)?;
+
+        Ok(token_data.claims)
+    }
+}
+
+/// Login request.
+#[derive(Debug, Clone, serde::Deserialize)]
+#[serde(tag = "method", content = "data")]
+pub enum AuthLoginRequest {
+    #[serde(rename = "siwe")]
+    Siwe { message: String, signature: String },
+
+    #[serde(rename = "std")]
+    Std { body: String, signature: String },
+}
+
+impl AuthLoginRequest {
+    /// Verify the login request and return the authenticated claims.
+    pub async fn verify(&self, domain: &str, provider: Address) -> Result<Claims, anyhow::Error> {
+        let address = match self {
+            Self::Siwe { message, signature } => {
+                let message: siwe::Message = message.parse().context("malformed SIWE message")?;
+                let signature: Vec<u8> =
+                    signature.from_hex().context("malformed SIWE signature")?;
+                let signature: [u8; 65] = signature
+                    .try_into()
+                    .map_err(|_| anyhow!("malformed SIWE signature"))?;
+
+                if !message.valid_now() {
+                    return Err(anyhow!("message is not yet valid or has expired"));
+                }
+
+                let expected_statement = format!("Provider: {}", provider.to_bech32());
+                if message.statement != Some(expected_statement) {
+                    return Err(anyhow!("message does not have the expected statement"));
+                }
+
+                let verification_opts = siwe::VerificationOpts {
+                    domain: Some(domain.parse()?),
+                    ..Default::default()
+                };
+                message.verify(&signature, &verification_opts).await?;
+
+                Address::from_eth(&message.address)
+            }
+            Self::Std { body, signature } => {
+                // Decode Base64-encoded body.
+                let raw_body = BASE64_STANDARD
+                    .decode(body)
+                    .context("malformed authentication statement body")?;
+                let body: StdAuthBody = cbor::from_slice(&raw_body)
+                    .context("malformed authentication statement body")?;
+                // Validate body.
+                body.validate(domain, provider)?;
+
+                // Decode Base64-encoded signature.
+                let signature = BASE64_STANDARD
+                    .decode(signature)
+                    .context("malformed signature")?
+                    .into();
+                // Verify signature.
+                let ctx = signature::context::get_chain_context_for(STD_AUTH_CONTEXT_BASE);
+                body.signer.verify(&ctx, &raw_body, &signature)?;
+
+                Address::from_sigspec(
+                    &SignatureAddressSpec::try_from_pk(&body.signer)
+                        .ok_or(anyhow!("unsupported public key scheme"))?,
+                )
+            }
+        };
+
```

Comment:
Should we require that the SIWE/StdAuthBody expiry time is short? (e.g. not_before/not_after times are close around the current time).

Reply:
Yeah, that would probably be unusual. I guess it's the client's responsibility not to sign messages with long validity. 👍